File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    n = input()
    if n == '.':
        break
    stack = []
    for i in n:
        if i in ['(', '[']:
            stack.append(i)
        elif i == ')':
            if stack and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)
        elif i == ']':
            if stack and stack[-1] == '[':
                stack.pop()
            else:
                stack.append(i)
        if i == '.':
            break
    if len(stack) == 0:
        print('yes')
    else:
        print('no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;


public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		while(true) {
			Stack<Character> stack = new Stack<>();
			String s = br.readLine();
			if(s.equals(".")) break;
			
			for(int i = 0; i<s.length(); i++) {
				char c = s.charAt(i);
				if(c == '(' || c == '[') {
					stack.push(c);
				}else if(c == ')') {
					if(!stack.isEmpty() && stack.peek() == '(') {
						stack.pop();
					}else {
						stack.push(c);
					}
				}else if(c == ']') {
					if(!stack.isEmpty() && stack.peek() == '[') {
						stack.pop();
					}else {
						stack.push(c);
					}
				}
			}
			if(stack.isEmpty()) {
				sb.append("yes").append('\n');
			}else {
				sb.append("no").append('\n');
			}
		}
		
		System.out.println(sb);
		
		
		
		
		
		
	}

}